### PR TITLE
fix(sqlparser): typed string must be single quoted

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

See #17461. Somehow a user wrote the invalid SQL `Date t::date` but it was interpreted by our sqlparser as `(Date 't')::date` and resulting in evaluation error.

The root cause is upstream https://github.com/sqlparser-rs/sqlparser-rs/pull/235 support of HiveSQL where an unquoted identifier can be treated as a literal string value. It is never the case for PostgreSQL.

After the fix, we report the following error message:
```
dev=> select date t::date from generate_series('2024-01-01 00:00:00+00'::timestamptz, now(), interval '1 month') as s(t);
ERROR:  Failed to run the query

Caused by:
  sql parser error: expected end of statement, found: ::
LINE 1: select date t::date from generate_series('2024-01-01 00:00:00+00'::timestamptz, now(), interval '1 month') as s(t);
                     ^
```
which is similar to PostgreSQL:
```
test=# select date t::date from generate_series('2024-01-01 00:00:00+00'::timestamptz, now(), interval '1 month') as s(t);
ERROR:  syntax error at or near "::"
LINE 1: select date t::date from generate_series('2024-01-01 00:00:0...
                     ^
```

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [x] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/sqlparser/tests/testdata/show.yaml [hunk 1]
```diff
@@ -11,7 +11,7 @@
 - input: SHOW TABLES FROM t
   formatted_sql: SHOW TABLES FROM t
   formatted_ast: 'ShowObjects { object: Table { schema: Some(Ident { value: "t", quote_style: None }) }, filter: None }'
-- input: SHOW TABLES FROM t LIKE "t%"
+- input: SHOW TABLES FROM t LIKE 't%'
```
Comment (xiangjinwu) [inline src/sqlparser/tests/testdata/show.yaml:14]: Strictly speaking this fix is a breaking change. But it was wrong syntax and our documentation has been using single quotes correctly. The impact shall be negligible.

**Comment by xiangjinwu** [comment]
https://github.com/risingwavelabs/risingwave/blob/18f99803f2719c047594a6a55e2abb5cfe82d3ec/e2e_test/udf/external_udf.slt#L8

This `as int_42` has been mis-parsed as `SingleQuotedDef`
https://github.com/risingwavelabs/risingwave/blob/18f99803f2719c047594a6a55e2abb5cfe82d3ec/src/sqlparser/src/ast/mod.rs#L2885-L2890

cc @wangrunji0408 Shall we introduce a new variant for (unquoted or double quoted) **identifier**?

**Review by lmatz — APPROVED** [review]

**Review by stdrc — APPROVED** [review]
LGTM!

**Comment by wangrunji0408** [comment]
> cc @wangrunji0408 Shall we introduce a new variant for (unquoted or double quoted) identifier?

added

**Review by wangrunji0408 — APPROVED** [review]

File: src/sqlparser/src/parser.rs [hunk 2]
```diff
@@ -3550,28 +3552,20 @@ impl Parser<'_> {
...
-                Ok(FunctionDefinition::DoubleDollarDef(value.value))
-            }
-            _ => Ok(FunctionDefinition::SingleQuotedDef(
-                self.parse_literal_string()?,
-            )),
-        }
+        alt((
+            single_quoted_string.map(FunctionDefinition::SingleQuotedDef),
+            dollar_quoted_string.map(FunctionDefinition::DoubleDollarDef),
+            Self::parse_identifier.map(|i| FunctionDefinition::Identifier(i.value)),
```
Comment (xiangjinwu) [inline src/sqlparser/src/parser.rs:3558]: When I said _(unquoted or double quoted) **identifier**_ I was thinking of `.real_value()` but now I realize `.value` has better backward compatibility:

When the SQL to parse is `as camelCase`, the old behavior gets `camelCase` but the identifier's `real_value` is `camelcase`. Although `as "camelCase"` is a case-sensitive identifier, I agree it is better to encourage `as 'camelCase'` if breaking change is inevitable.

**Pull request merged**